relayer/src/keyring/store.rs

```diff
@@ -15,137 +12,200 @@ use bitcoin::{
-                    Kind::KeyStoreOperation
-                        .context(format!("failed to create keys folder: {:?}", e))
+                let keys_folder = disk_store_path(chain_config.id.as_str()).map_err(|e| {
+                    Kind::KeyStore.context(format!("failed to create keys folder: {:?}", e))
                 })?;
+
                 fs::create_dir_all(keys_folder.clone())
-                    .map_err(|_| Kind::KeyStoreOperation.context("error creating keys folder"))?;
+                    .map_err(|_| Kind::KeyStore.context("error creating keys folder"))?;
 
-                Ok(KeyRing::TestKeyStore {
-                    store: Box::<Path>::from(keys_folder),
-                    chain_config,
-                })
+                Ok(Self::Disk(Disk::new(
+                    chain_config.key_name,
+                    chain_config.account_prefix,
+                    keys_folder,
+                )))
             }
         }
     }
 
+    pub fn get_key(&self) -> Result<KeyEntry, Error> {
+        match self {
+            KeyRing::Memory(m) => m.get_key(),
+            KeyRing::Disk(d) => d.get_key(),
+        }
+    }
+
+    pub fn add_key(&mut self, key_entry: KeyEntry) -> Result<(), Error> {
+        match self {
+            KeyRing::Memory(m) => m.add_key(key_entry),
+            KeyRing::Disk(d) => d.add_key(key_entry),
+        }
+    }
+
     /// Get key from seed file
-    fn key_from_seed_file(
-        &self,
-        key_file_content: &str,
-        chain_config: &ChainConfig,
-    ) -> Result<KeyEntry, Error> {
-        let key_json: Value =
+    pub fn key_from_seed_file(&self, key_file_content: &str) -> Result<KeyEntry, Error> {
+        #[derive(Serialize, Deserialize)]
+        struct KeyFile {
+            name: String,
+            r#type: String,
+            address: String,
+            pubkey: String,
+            mnemonic: String,
+        }
+
+        let key_file: KeyFile =
             serde_json::from_str(key_file_content).map_err(|e| Kind::InvalidKey.context(e))?;
 
-        let key: KeyEntry;
-
-        let _mnemonic: String = "".to_string();
-        let mnemonic_value = key_json.get("mnemonic");
-        match mnemonic_value {
-            Some(m) => {
-                let mnemonic = m.as_str();
-                match mnemonic {
-                    Some(v) => {
-                        key = self
-                            .key_from_mnemonic(v, chain_config)
-                            .map_err(|e| Kind::InvalidMnemonic.context(e))?;
-                        Ok(key)
-                    }
-                    None => Err(Kind::InvalidMnemonic
-                        .context("invalid key file, cannot find mnemonic".to_string())
-                        .into()),
-                }
-            }
-            None => Err(Kind::InvalidMnemonic
-                .context("invalid key file, cannot find mnemonic".to_string())
-                .into()),
-        }
+        let mut key = self
```

Comment:
Is the account prefix encoded in the public key as well? As far as I can tell only the address itself can be derived from the public key, hence why we are using the configured account prefix. If that's the case, I guess we'd have to ask for the account prefix on the command line when importing a key from the mnemonic?

As for importing from the key file, the best solution imho would be to:
- use the address and pub key given in the key file
- extract the account from the address given in the key file
- extract the private key from the mnemonic
- get the public key from the private key, ensure it matches the one in the key file

What do you think?

Reply:
I think the solution looks good. For the question related to importing a key from the mnemonic, if you have the account_prefix specified in the config, the relayer can read from that since you have to pass a chain-id. That's how the go relayer does it, then you can generate the address based on BIP-44. You can see how it works in here:
https://docs.cosmos.network/v0.42/basics/accounts.html

Reply:
Alright, so if we allow adding a key from the mnemonic alone we'll use the configured account prefix, and when we import a key from a key file, we'll use the embedded account, per the algorithm above.